packages/core/src/codewhisperer/util/authUtil.ts

```diff
@@ -418,55 +418,59 @@ export class AuthUtil {
     public isValidCodeTransformationAuthUser(): boolean {
         return this.isEnterpriseSsoInUse() && this.isConnectionValid()
     }
-}
-
-/**
- * Returns a snapshot of the overall auth state of
- * CodeWhisperer + Chat features.
- */
-export async function getChatAuthState(cwAuth = AuthUtil.instance): Promise<FeatureAuthState> {
-    const currentConnection = cwAuth.conn
-
-    if (currentConnection === undefined) {
-        return buildFeatureAuthState(AuthStates.disconnected)
-    }
-    if (!isSsoConnection(currentConnection)) {
-        throw new ToolkitError(`Connection "${currentConnection.id}" is not a valid type: ${currentConnection.type}`)
-    }
 
-    // The state of the connection may not have been properly validated
-    // and the current state we see may be stale, so refresh for latest state.
-    await cwAuth.auth.refreshConnectionState(currentConnection)
+    /**
+     * Returns a snapshot of the overall auth state of
+     * CodeWhisperer + Chat features.
+     */
+    public async getChatAuthState(shouldRefresh: boolean = true): Promise<FeatureAuthState> {
+        const currentConnection = this.conn
 
-    // default to expired to indicate reauth is needed if unmodified
-    const state: FeatureAuthState = buildFeatureAuthState(AuthStates.expired)
+        if (currentConnection === undefined) {
+            return buildFeatureAuthState(AuthStates.disconnected)
+        }
+        if (!isSsoConnection(currentConnection)) {
+            throw new ToolkitError(
+                `Connection "${currentConnection.id}" is not a valid type: ${currentConnection.type}`
+            )
+        }
 
-    if (isBuilderIdConnection(currentConnection)) {
-        // Regardless, if using Builder ID, Amazon Q is unsupported
-        state[Features.amazonQ] = AuthStates.unsupported
-    }
+        // The state of the connection may not have been properly validated
+        // and the current state we see may be stale, so refresh for latest state.
+        if (shouldRefresh) {
+            await this.auth.refreshConnectionState(currentConnection)
```

Comment:
If this refresh emit connection state change event, can it run into a infinite loop?

Reply:
Example: If it refreshConnectionState ---> emit connection changed event ---> connection change hook ---> `refreshConnectionState`

Reply:
no it should converge, because each time it is called it compares to old values.